compiler/qsc_eval/src/intrinsic/utils.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::collections::hash_map::Entry;
+
+use num_bigint::BigUint;
+use num_complex::{Complex, Complex64};
+use num_traits::{One, Zero};
+use rustc_hash::FxHashMap;
+
+pub fn split_state(
+    qubits: &[usize],
+    state: Vec<(BigUint, Complex64)>,
+    qubit_count: usize,
+) -> Option<Vec<(BigUint, Complex64)>> {
+    let state = state.into_iter().collect::<FxHashMap<_, _>>();
+    let mut dump_state = FxHashMap::default();
+    let mut other_state = FxHashMap::default();
+
+    // Compute the mask for the qubits to dump and the mask for the other qubits.
+    let mut dump_mask = BigUint::zero();
+    let mut other_mask = BigUint::zero();
+    for q in 0..qubit_count {
+        // Note that the qubit order is reversed to match the order of the qubits in the state.
+        if qubits.contains(&q) {
+            dump_mask.set_bit((qubit_count - q - 1) as u64, true);
+        } else {
+            other_mask.set_bit((qubit_count - q - 1) as u64, true);
+        }
+    }
+
+    // Try to split out the state for the given qubits from the whole state, detecting any entanglement
+    // and returning `None` if the qubits are not separable.
+    let mut state_iter = state.iter();
+    let (base_label, base_val) = state_iter.next().expect("state should never be empty");
+    let dump_base_label = base_label & &dump_mask;
+    let other_base_label = base_label & &other_mask;
+    let mut dump_norm = 1.0_f64;
+    dump_state.insert(dump_base_label.clone(), Complex64::one());
+    other_state.insert(other_base_label.clone(), Complex64::one());
+    for (curr_label, curr_val) in state_iter {
+        let dump_label = curr_label & &dump_mask;
+        let other_label = curr_label & &other_mask;
+
+        // If either the state identified by the dump mask or the state identified by the other mask
+        // is None, that means it has zero amplitude and we can conclude the state is not separable.
+        let Some(dump_val) = state.get(&(&dump_label | &other_base_label)) else {
+            return None;
+        };
+        let Some(other_val) = state.get(&(&dump_base_label | &other_label)) else {
+            return None;
+        };
+
+        if !(dump_val * other_val - base_val * curr_val)
+            .norm()
+            .is_nearly_zero()
+        {
+            // Coefficients are not equal, so the state is not separable.
+            return None;
+        }
+
+        if let Entry::Vacant(entry) = dump_state.entry(dump_label) {
+            let amplitude = curr_val / other_val;
```

Comment:
I'm not sure I understand why the amplitude is the division of the current label value over the other value.

Reply:
I added a comment to try and clarify. The amplitude we are capturing here is intermingled in the original state vector, so we calculated the amplitude of the "other" state and divide that out to get the amplitude of just the desired qubits.